lzy/graph-executor-2/src/main/java/ai/lzy/graph/algo/Algorithms.java

```diff
@@ -0,0 +1,110 @@
+                );
+            }
+            taskMap.put(task.id(), task);
+            buildChannelsForTask(task, channelInputs, channelOutputs);
+        }
+
+        for (String channelId: channels) {
+            for (TaskState input: channelInputs.get(channelId)) {
+                for (TaskState output: channelOutputs.get(channelId)) {
+                    input.tasksDependedFrom().add(output.id());
+                    output.tasksDependedOn().add(input.id());
+                }
+            }
+        }
+
+        if (isCyclic(taskMap)) {
+            throw new IllegalStateException(
+                "Graph with id <%s> contains cycles".formatted(graph.id())
+            );
+        }
+    }
+
+    private static void buildChannelsForTask(TaskState task,
+                                             Map<String, List<TaskState>> channelInputs,
+                                             Map<String, List<TaskState>> channelOutputs)
+    {
+        final List<String> inputs = task.taskSlotDescription().slots().stream()
+            .filter(s -> s.direction() == TaskSlotDescription.Slot.Direction.INPUT)
+            .map(TaskSlotDescription.Slot::name)
+            .toList();
+
+        final List<String> outputs = task.taskSlotDescription().slots().stream()
+            .filter(s -> s.direction() == TaskSlotDescription.Slot.Direction.OUTPUT)
+            .map(TaskSlotDescription.Slot::name)
+            .toList();
+
+        final Set<String> intersection = new HashSet<>(inputs);
+        intersection.retainAll(outputs);
+
+        if (intersection.size() > 0) {
+            throw new IllegalStateException(
+                "Slots %s of task <%s> are inputs and outputs in the same time"
+                    .formatted(Arrays.toString(intersection.toArray()), task.id())
+            );
+        }
+
+        task.taskSlotDescription().slotsToChannelsAssignments().entrySet().stream()
+            .filter(e -> outputs.contains(e.getKey()))
+            .map(e -> channelInputs.computeIfAbsent(e.getValue(), k -> new ArrayList<>()))
+            .forEach(l -> l.add(task));
+        task.taskSlotDescription().slotsToChannelsAssignments().entrySet().stream()
+            .filter(e -> inputs.contains(e.getKey()))
+            .map(e -> channelOutputs.computeIfAbsent(e.getValue(), k -> new ArrayList<>()))
+            .forEach(l -> l.add(task));
+    }
+
+    private static boolean isCyclic(Map<String, TaskState> taskMap) {
+        Set<TaskState> visited = new HashSet<>();
+        Set<TaskState> stack = new HashSet<>();
+
+        for (TaskState vertex : taskMap.values()) {
+            if (isCyclicUtil(taskMap, vertex, visited, stack)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean isCyclicUtil(Map<String, TaskState> taskMap, TaskState task,
+                                        Set<TaskState> visited, Set<TaskState> stack)
+    {
+        if (stack.contains(task)) {
```

Comment:
you may wish to replace double lookup by replacing the `.contains()` call with `boolean added = stack.add(task)`